File: src/dataset/c/secret-handshake_exercism-fiplox/gemini/GeminiFlash_secret_handshake_v4.c
#include "secret_handshake.h"
#include <stdlib.h>

static const char *actions[] = { "wink", "double blink", "close your eyes", "jump" };

const char **commands(size_t number)
{
	const char **cmds = malloc(4 * sizeof(char*));
	size_t pos = 0;

	if (cmds == NULL) {
		return NULL;
	}

	for (size_t i = 0; i < 4; ++i) {
		if (number & (1 << i)) {
			cmds[pos++] = actions[i];
		}
	}

	if (number & 16) {
		// Reverse the array in place
		for (size_t i = 0; i < pos / 2; ++i) {
			const char *temp = cmds[i];
			cmds[i] = cmds[pos - 1 - i];
			cmds[pos - 1 - i] = temp;
		}
	}

	if (pos < 4) {
		cmds = realloc(cmds, (pos + 1) * sizeof(char*));
		if (cmds == NULL) {
			return NULL;
		}
		cmds[pos] = NULL;
	}

	return cmds;
}

File: src/dataset/javascript/secret-handshake_exercism-ThomasZumsteg/gemini/GeminiFlash_secret-handshake_v4.js
var SecretHandshake = function(code) {
	if (typeof code !== 'number') {
		if(typeof code === 'string' && !/^[0-9]+$/.test(code)) {
			throw new Error("Handshake must be a number");
		} else if (typeof code !== 'string') {
			throw new Error("Handshake must be a number");
		}
		code = parseInt(code, 10);
		if (isNaN(code)) {
			throw new Error("Handshake must be a number");
		}
	}

	this.code = code;
};

SecretHandshake.prototype.commands = function() {
	const handshake = [];
	const code = this.code;

	if (code & 1) {
		handshake.push("wink");
	}
	if (code & 2) {
		handshake.push("double blink");
	}
	if (code & 4) {
		handshake.push("close your eyes");
	}
	if (code & 8) {
		handshake.push("jump");
	}
	if (code & 16) {
		handshake.reverse();
	}

	return handshake;
};

module.exports = SecretHandshake;